homeassistant/components/evohome.py

```diff
@@ -0,0 +1,606 @@
+# current state - the exception is when zones pull their own schedules.
+# However, any entity can call methods that will (eventually) change state.
+PARALLEL_UPDATES = 0
+
+# these are specific to this component
+DOMAIN = 'evohome'
+DATA_EVOHOME = 'data_' + DOMAIN
+DISPATCHER_EVOHOME = 'dispatcher_' + DOMAIN
+
+MIN_TEMP = 5
+MAX_TEMP = 35
+SCAN_INTERVAL_DEFAULT = 180
+SCAN_INTERVAL_MAX = 300
+
+CONF_LOCATION_IDX = 'location_idx'
+
+# Validation of the user's configuration.
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_LOCATION_IDX, default=0): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+# these are for the controller's opmode/state and the zone's state
+EVO_RESET = 'AutoWithReset'
+EVO_AUTO = 'Auto'
+EVO_AUTOECO = 'AutoWithEco'
+EVO_AWAY = 'Away'
+EVO_DAYOFF = 'DayOff'
+EVO_CUSTOM = 'Custom'
+EVO_HEATOFF = 'HeatingOff'
+
+TCS_MODES = [
+    EVO_RESET,
+    EVO_AUTO,
+    EVO_AUTOECO,
+    EVO_AWAY,
+    EVO_DAYOFF,
+    EVO_CUSTOM,
+    EVO_HEATOFF
+]
+
+# bit masks for dispatcher packets
+EVO_MASTER = 0x01
+EVO_SLAVE = 0x02
+
+# these are used to help prevent E501 (line too long) violations
+GWS = 'gateways'
+TCS = 'temperatureControlSystems'
+
+
+def setup(hass, config):
+    """Create a Honeywell (EMEA/EU) evohome CH/DHW system.
+
+    One controller with 0+ heating zones (e.g. TRVs, relays) and, optionally, a
+    DHW controller.  Does not work for US-based systems.
+    """
+    # Used for internal data, such as installation, state & timers...
+    hass.data[DATA_EVOHOME] = {}
+
+    domain_data = hass.data[DATA_EVOHOME]
+    domain_data['timers'] = {}
+
+    # Pull the configuration parameters...
+    domain_data['params'] = dict(config[DOMAIN])
+    # scan_interval - start with default value
+    domain_data['params'][CONF_SCAN_INTERVAL] = SCAN_INTERVAL_DEFAULT
+
+    if _LOGGER.isEnabledFor(logging.DEBUG):  # then redact username, password
+        tmp = dict(domain_data['params'])
+        tmp[CONF_USERNAME] = 'REDACTED'
+        tmp[CONF_PASSWORD] = 'REDACTED'
+
+        _LOGGER.debug("setup(): Configuration parameters: %s", tmp)
+
+    from evohomeclient2 import EvohomeClient
+
+    _LOGGER.debug("setup(): API call [4 request(s)]: client.__init__()...")
+
+    try:
+        # there's a bug in evohomeclient2 v0.2.7: the client.__init__() sets
+        # the root loglevel (debug=?), so must remember it now...
+        log_level = logging.getLogger().getEffectiveLevel()
+
+        client = EvohomeClient(
+            domain_data['params'][CONF_USERNAME],
+            domain_data['params'][CONF_PASSWORD],
+            debug=False
+        )
+        # ...then restore it to what it was before instantiating the client
+        logging.getLogger().setLevel(log_level)
+
+    except requests.RequestException as err:
+        if str(HTTP_BAD_REQUEST) in str(err):
```

Comment:
this feels weird, better to check `if err.response and err.response.status_code == 400:`

Reply:
Ha! The main reason for this is that I didn't know any Python before 3mo ago (and barely any programming).  I will fix it.

Reply:
sorted